### ISSUE TITLE
Phase 27.3 — FewShotAdapter: Meta-Learning & Rapid Task Adaptation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#596]
## Overview

`FewShotAdapter` enables rapid adaptation to new tasks and domains from minimal examples — the "learning to learn" capability. Instead of training from scratch, it leverages meta-learned representations to achieve good performance with as few as 1–5 examples.

**Theoretical basis**: Finn et al.'s MAML (2017) — learn an initialization that is easy to fine-tune; Snell et al.'s Prototypical Networks (2017) — classify by distance to class prototypes in embedding space; Vinyals et al.'s Matching Networks (2016) — attention over support set.

### `TaskEmbedding` (frozen dataclass)

```python
@dataclass(frozen=True, slots=True)
class TaskEmbedding:
    task_id: str
    embedding: tuple[float, ...]            # task representation vector
    support_set: tuple[str, ...]            # few-shot examples (instance IDs)
    query_set: tuple[str, ...] = ()         # evaluation examples
    domain: str | None = None               # source domain identifier
    difficulty: float = 0.5                 # estimated task difficulty [0,1]
    similarity_to_known: float = 0.0        # max similarity to previously seen tasks
```

### `AdaptationResult` (frozen dataclass)

```python
@dataclass(frozen=True, slots=True)
class AdaptationResult:
    task_id: str
    adapted_params: Mapping[str, Any]       # adapted model parameters/weights
    performance: float                      # [0.0, 1.0] on query set
    shots_used: int                         # number of examples consumed
    confidence: float                       # adaptation confidence [0,1]
    adaptation_time_ms: float               # wall-clock adaptation time
    transfer_source: str | None = None      # domain that provided the best transfer
    abstraction_level: str | None = None    # level at which transfer operated
```

## Protocol

```python
@runtime_checkable
class FewShotAdapter(Protocol):
    async def adapt(
        self, task: TaskEmbedding, max_shots: int = 5
    ) -> AdaptationResult: ...

    async def evaluate(
        self, result: AdaptationResult, test_instances: Sequence[str]
    ) -> float: ...

    async def get_task_embedding(
        self, task_description: str, examples: Sequence[str]
    ) -> TaskEmbedding: ...

    async def select_support_set(
        self, task: TaskEmbedding, candidates: Sequence[str], k: int = 5
    ) -> Sequence[str]: ...
```

## Implementation: `MetaFewShotAdapter`

**Algorithm** — Gradient-Free Meta-Adaptation:
1. **Embed task**: Compute task embedding from description + examples using compositional encoder
2. **Retrieve prototypes**: Find nearest task prototypes from meta-knowledge base
3. **Compute fast weights**: Linear combination of prototype parameters weighted by task similarity
4. **Adapt**: Fine-tune fast weights on support set using prototype refinement (no gradients)
5. **Evaluate**: Score adapted model on query set, update meta-knowledge if performance > threshold

**Key features**:
- `_embed_task(description, examples) -> tuple[float, ...]` — compositional task encoding
- `_find_prototypes(embedding, k) -> Sequence[Prototype]` — nearest-neighbor prototype retrieval
- `_compute_fast_weights(prototypes, similarities) -> Mapping` — similarity-weighted parameter combination
- `_refine_on_support(weights, support_set) -> Mapping` — iterative prototype refinement (no backprop)
- `_select_informative_examples(candidates, embedding) -> Sequence` — maximum diversity + relevance selection
- Meta-knowledge base: `dict[str, Prototype]` with periodic consolidation via MemoryConsolidator (18.2)
- `asyncio.Lock` for meta-knowledge updates

## Null Implementation & Factory

```python
class NullFewShotAdapter(FewShotAdapter):
    async def adapt(self, task, max_shots=5):
        return AdaptationResult(task_id=task.task_id, adapted_params={}, performance=0.0, shots_used=0, confidence=0.0, adaptation_time_ms=0.0)
    # ...

def make_few_shot_adapter(*, strategy: str = "meta", prototype_k: int = 10) -> FewShotAdapter:
    match strategy:
        case "meta": return MetaFewShotAdapter(prototype_k=prototype_k)
        case "null": return NullFewShotAdapter()
        case _: raise ValueError(f"Unknown strategy: {strategy}")
```

## Metrics (Prometheus)

| Metric | Type | Description |
|--------|------|-------------|
| `asi_few_shot_adaptations_total` | Counter | Total adaptation attempts |
| `asi_few_shot_adaptation_duration_seconds` | Histogram | Time per adaptation |
| `asi_few_shot_performance` | Histogram | Post-adaptation performance distribution |
| `asi_few_shot_shots_used` | Histogram | Number of examples used per adaptation |
| `asi_few_shot_meta_knowledge_size` | Gauge | Number of prototypes in meta-knowledge base |

## Integration Points

- **DomainMapper (27.1)** — domain mappings guide which prototypes are relevant
- **AbstractionEngine (27.2)** — abstraction level determines adaptation granularity
- **MemoryConsolidator (18.2)** — prototype consolidation and episodic-to-semantic compression
- **CurriculumDesigner (27.4)** — adaptation results inform curriculum difficulty calibration

## Test Targets (12)

1. `test_task_embedding_immutability`
2. `test_adaptation_result_bounds`
3. `test_adapt_known_task_high_performance`
4. `test_adapt_novel_task_uses_prototypes`
5. `test_adapt_respects_max_shots`
6. `test_select_support_set_diversity`
7. `test_get_task_embedding_deterministic`
8. `test_evaluate_on_test_instances`
9. `test_meta_knowledge_grows_on_success`
10. `test_null_adapter_returns_defaults`
11. `test_factory_meta`
12. `test_factory_unknown_raises`

---
*Phase 27.3 gives the system the ability to learn rapidly from minimal data — a hallmark of human-level intelligence.*
*References: [#593](https://github.com/web3guru888/asi-build/discussions/593) (Phase 27 Planning) | [#594](https://github.com/web3guru888/asi-build/issues/594) (27.1) | [#595](https://github.com/web3guru888/asi-build/issues/595) (27.2)*

**Comment by web3guru888** [comment]
## Implementation Notes

### Gradient-Free Fast Weight Computation

```python
def _compute_fast_weights(
    self, prototypes: Sequence[Prototype], similarities: Sequence[float]
) -> Mapping[str, Any]:
    """Similarity-weighted combination of prototype parameters."""
    # Softmax normalization
    exp_sims = [math.exp(s * self._temperature) for s in similarities]
    total = sum(exp_sims)
    weights = [e / total for e in exp_sims]
    
    # Weighted combination
    result = {}
    for key in prototypes[0].params:
        result[key] = sum(
            w * p.params[key] for w, p in zip(weights, prototypes)
        )
    return result
```

### Support Set Selection (Greedy DPP Approximation)

```python
async def select_support_set(
    self, task: TaskEmbedding, candidates: Sequence[str], k: int = 5
) -> Sequence[str]:
    # 1. Filter by relevance
    relevant = [
        (c, self._relevance(task.embedding, c))
        for c in candidates
        if self._relevance(task.embedding, c) >= 0.3
    ]
    
    # 2. Greedy diversification
    selected = [max(relevant, key=lambda x: x[1])]
    for _ in range(k - 1):
        best = max(
            [c for c in relevant if c not in selected],
            key=lambda c: c[1] * min(
                self._distance(c[0], s[0]) for s in selected
            ),
        )
        selected.append(best)
    
    return [s[0] for s in selected]
```

### Prototype Refinement (No Backprop)

```python
def _refine_on_support(
    self, weights: Mapping[str, Any], support_set: Sequence[str]
) -> Mapping[str, Any]:
    """Iterative prototype refinement without gradient computation."""
    refined = dict(weights)
    for example in support_set:
        embedding = self._embed_instance(example)
        # Move prototype toward example (online centroid update)
        for key in refined:
            refined[key] = (1 - self._learning_rate) * refined[key] + \
                           self._learning_rate * embedding.get(key, 0.0)
    return refined
```

### Mypy Narrowing Table

| Variable | Narrowed From | Narrowed To | Guard |
|----------|--------------|-------------|-------|
| `domain` | `str \| None` | `str` | `if task.domain is not None` |
| `query_set` | `tuple[str,...] \| ()` | `tuple[str,...]` | `if task.query_set` |
| `transfer_source` | `str \| None` | `str` | `if result.transfer_source is not None` |

### Test Skeleton

```python
@pytest.mark.asyncio
async def test_adapt_known_task_high_performance():
    adapter = make_few_shot_adapter(strategy="meta")
    # Pre-populate meta-knowledge with a known task
    known_task = TaskEmbedding("known", (0.5,) * 64, ("ex1", "ex2"), ("q1",))
    result = await adapter.adapt(known_task, max_shots=3)
    assert result.shots_used <= 3
    assert result.confidence >= 0.0

@pytest.mark.asyncio
async def test_select_support_set_diversity():
    adapter = make_few_shot_adapter(strategy="meta")
    task = TaskEmbedding("t1", (1.0,) * 64, (), ())
    candidates = [f"c{i}" for i in range(20)]
    selected = await adapter.select_support_set(task, candidates, k=5)
    assert len(selected) == 5
    assert len(set(selected)) == 5  # all unique
```

### Implementation Order (14 steps)

1. `TaskEmbedding` frozen dataclass
2. `AdaptationResult` frozen dataclass
3. `FewShotAdapter` Protocol with `@runtime_checkable`
4. `Prototype` internal dataclass (task_id, embedding, params, performance)
5. `_embed_task` compositional encoder
6. `_find_prototypes` nearest-neighbor retrieval
7. `_compute_fast_weights` softmax-weighted combination
8. `_refine_on_support` online centroid update
9. `_relevance` task-candidate scoring
10. `select_support_set` greedy DPP
11. `MetaFewShotAdapter.adapt` full pipeline
12. `evaluate` test set scoring
13. `NullFewShotAdapter`
14. `make_few_shot_adapter` factory

*S&T: [#603](https://github.com/web3guru888/asi-build/discussions/603) | Q&A: [#604](https://github.com/web3guru888/asi-build/discussions/604)*
